File: cpp/778_水位上升的泳池中游泳_medium.cpp
class Dist {
public:
    int h, x, y;
    Dist(int _h, int _x, int _y): h(_h), x(_x), y(_y) {}
    bool operator < (const Dist &that) const {
        return h > that.h;
    }
};
class Solution {
public:
    int swimInWater(vector<vector<int>>& grid) {
        int rows = grid.size();
        int columns = grid[0].size();
        vector<bool> seen(rows*columns);
        int ans = 0;
        vector<vector<int>> directions = {
            {0, 1},
            {1, 0},
            {0, -1},
            {-1, 0}
        };
        priority_queue <Dist> q;
        q.emplace(grid[0][0], 0, 0);
        while (!q.empty()) {
            auto [t, x, y] = q.top();
            q.pop();
            int id = x*columns+y;
            if (seen[id])
                continue;
            ans = max(ans, t);
            if (x == rows-1 && y == columns-1)
                break;
            seen[id] = true;
            for (int i=0; i<4; i++) {
                int nx = x + directions[i][0];
                int ny = y + directions[i][1];
                if (nx>=0&&nx<rows && ny>=0&&ny<columns && !seen[nx*columns+ny])
                    q.emplace(grid[nx][ny], nx, ny);
            }
        }
        return ans;
    }
};

File: python/778_水位上升的泳池中游泳_medium.py
class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        rows, columns = len(grid), len(grid[0])
        seen = set()
        q = [(grid[0][0], 0, 0)]
        directions = [(-1, 0), (0, -1), (1, 0), (0, 1)]
        ans = 0
        while q:
            t, x, y = heapq.heappop(q)
            idx = x*columns+y
            if idx in seen:
                continue
            ans = max(ans, t)
            if x == rows-1 and y == columns-1:
                break
            seen.add(idx)
            for dx, dy in directions:
                nx, ny = x + dx, y + dy
                if 0<=nx<rows and 0<=ny<columns and nx*columns+ny not in seen:
                    heapq.heappush(q, (grid[nx][ny], nx, ny))
        return ans
            